File: examples/cpp/quick_start/quick_start_simulator.cpp
/**
 * Part of the openDAQ stand-alone application quick start guide. Starts
 * an openDAQ server on localhost that contains a simulated device.
 */

#include <opendaq/opendaq.h>
#include <iostream>

using namespace daq;

int main(int /*argc*/, const char* /*argv*/[])
{
    using namespace std::chrono_literals;

    StringPtr loggerPath = "ref_device_simulator.log";

    PropertyObjectPtr config = PropertyObject();
    config.addProperty(StringProperty("Name", "Reference device simulator"));
    config.addProperty(StringProperty("LocalId", "RefDevSimulator"));
    config.addProperty(StringProperty("SerialNumber", "sim01"));
    config.addProperty(BoolProperty("EnableLogging", true));
    config.addProperty(StringProperty("LoggingPath", loggerPath));

    const InstanceBuilderPtr instanceBuilder = InstanceBuilder();
    instanceBuilder.addModulePath(MODULE_PATH);
    instanceBuilder.addDiscoveryServer("mdns");
    instanceBuilder.setRootDevice("daqref://device1", config);
    instanceBuilder.addLoggerSink(BasicFileLoggerSink(loggerPath));
    for (const auto& sink : DefaultSinks())
        instanceBuilder.addLoggerSink(sink);

    const InstancePtr instance = instanceBuilder.build();

    const auto servers = instance.addStandardServers();

    for (const auto& server : servers)
        server.enableDiscovery();
        
    std::cout << "Press \"enter\" to exit the application..." << std::endl;
    std::cin.get();
    return 0;
}

File: examples/dotnet/quick_start/quick_start_simulator.cs
/**
 * Part of the openDAQ stand-alone application quick start guide. Starts
 * an openDAQ server on localhost that contains a simulated device.
 */

using Daq.Core.Types;
using Daq.Core.Objects;
using Daq.Core.OpenDAQ;

// Create a simulation-device configuration object
PropertyObject config = CoreObjectsFactory.CreatePropertyObject();
config.AddProperty(CoreObjectsFactory.CreateStringProperty("Name", "Reference device simulator", visible: true));
config.AddProperty(CoreObjectsFactory.CreateStringProperty("LocalId", "RefDevSimulator", visible: true));
config.AddProperty(CoreObjectsFactory.CreateStringProperty("SerialNumber", "sim01", visible: true));

// Create an openDAQ(TM) instance builder to configure the instance
var instanceBuilder = OpenDAQFactory.InstanceBuilder();

// Add mDNS discovery service available for servers
instanceBuilder.AddDiscoveryServer("mdns");

// Set reference device as a root
instanceBuilder.SetRootDevice("daqref://device0", config);

// Creating a new instance from builder
var instance = instanceBuilder.Build();

// Start the openDAQ OPC UA and native streaming servers
var servers = instance.AddStandardServers();

foreach (var server in servers)
    server.EnableDiscovery();

Console.WriteLine();
Console.Write("Press a key to exit the application ...");
Console.ReadKey(intercept: true);
Console.WriteLine();
